certificate-manager/src/main/java/io/strimzi/certs/IpAndDnsValidation.java

```diff
@@ -0,0 +1,130 @@
+
+    // Pattern used to normalize the segments of the IPv6 address
+    private static final Pattern IPV6_LEADING_ZEROS_TRIMMING = Pattern.compile("(^|:)0+([\\da-f]+)(:|$)");
+
+    /**
+     * Returns true if the name is a valid DNS name. That means that it matches the DNS_NAME pattern
+     *
+     * @param name   Name which should be validated
+     *
+     * @return  True if the name is valid DNS name. False otherwise.
+     */
+    public static boolean isValidDnsName(String name) {
+        return name.length() <= 255
+                && (DNS_NAME.matcher(name).matches()
+                || (name.startsWith("*.") && DNS_NAME.matcher(name.substring(2)).matches()));
+    }
+
+    /**
+     * Returns true if the name is a valid IPv4 or IPv6 address.
+     *
+     * @param ip    IP address which should be validated
+     *
+     * @return  True if the name is valid IP address. False otherwise.
+     */
+    public static boolean isValidIpAddress(String ip) {
+        return isValidIpv4Address(ip) || isValidIpv6Address(ip);
+    }
+
+    /**
+     * Returns true if the name is a valid IPv4 address. Valid IPv4 address is an address which matches the IPV6_ADDRESS
+     * or IPV6_COMPRESSED_ADDRESS pattern.
+     *
+     * @param ip    IP address which should be validated
+     *
+     * @return  True if the name is valid IPv6 address. False otherwise.
+     */
+    public static boolean isValidIpv4Address(String ip) {
+        boolean matches = IPV4_ADDRESS.matcher(ip).matches();
+        if (matches) {
+            String[] split = ip.split("\\.");
+            for (String num : split) {
+                int i = Integer.parseInt(num);
+                if (i > 255) {
+                    return false;
+                }
+            }
+        }
+        return matches;
+    }
+
+    /**
+     * Returns true if the name is a valid IPv6 address. Valid IPv6 address is an address which matches the IPV4_ADDRESS
+     * pattern and each segment is equal or less than 255.
+     *
+     * @param ip    IP address which should be validated
+     *
+     * @return  True if the name is valid IPv4 address. False otherwise.
+     */
+    public static boolean isValidIpv6Address(String ip) {
+        return IPV6_ADDRESS.matcher(ip).matches() || IPV6_COMPRESSED_ADDRESS.matcher(ip).matches();
+    }
+
+    /**
+     * Normalizes the IPv6 address. Normalized address in Strimzi means that the IPv6 address is:
+     *     - uncompressed
+     *     - without leading 0
+     *     - lowercase
+     * This is based on the OpenSSL output. OpenSSL will normalize the IPv6 addresses in SANs this way. And we need to
+     * be able to compare the SANs to know when the SANs changed and a new cert should be generated.
+     *
+     * @param ip    IPv6 address which should be normalized
+     *
+     * @return  Normalized IPv6 address
+     */
+    public static String normalizeIpv6Address(String ip)   {
```

Comment:
Is it a precondition of this method that `ip` has been validated already? If so we should document that (and maybe call `ip` `validId` instead). Else perhaps we should check it here?